### DISCUSSION TITLE
DDR Dawn energy boosts ahead of night battles

### DISCUSSION THREAD

**Discussion by FrozenEmpire** [ChilliHugger/The-Lords-Of-Midnight#190] [Code Algorithm Questions] [answered]
Hi Chris,

I noticed that StartDawn is called at the beginning of night processing.

Increasing energies through resting, ahead of night battles, seems ok to me, however the dawn energy boosts also get applied here ahead of the nights battles... is that as per original DDR ?

```
void ddr_character::StartDawn ( void )
{
    if ( !IsDead() ) {        
        if ( IsResting() )
            IncreaseEnergy(RESTING_ENERGY_BOOST);        
        if ( IsAIControlled() )
            IncreaseEnergy( DAWN_ENERGY_BOOST_AI );
        else
            IncreaseEnergy( DAWN_ENERGY_BOOST );
    }
}
```

Regards,
Andrew

**Comment by IcemarkUK** [comment, marked as answer]
Short answer: Yes this is correct. The difference over the original code is that in the original the resting affect is applied immediately that you choose the command. Whereas for me, I set a flag and apply it here.

HOWEVER - I think I do have something wrong... But I am not sure if it's the difference in how I use energy over the original. Basically I think the 10 or 20 for AI non AI might be flipped. I need to investigate much deeper! Because that suggests that characters not loyal to the Moonprince get an extra 10 points of energy every night. Thinking about it, that kinda makes sense. I think the confusion may come from me originally labelling the variable as Tiredness rather than Energy, and then using terms like Reduce and not Increase.

But definitely wrong is that RESTING_ENERGY_BOOST = 32 which means the character always gets a full days rest. But they shouldn't  - they should get the amount of hours left in the day.

```
Night:                                              ;L6600
        CALL   CheckEndGameConditions
        CALL   ClearScreenBlack
        CALL   DisplayNightMessage
        CALL   FlushPrintBuffer
        CALL   ClearMapOfArmys
    
        XOR    AL,AL
L6610:  MOV    BYTE PTR CharacterNumber,AL
        MOV    BYTE PTR ArmyNumber,AL
        PUSH   AX
        PUSHF  
        CALL   GetCharReckless                      ;
        CMP    AL,DEAD                              ; Is character Dead
        JZ     L6652                                ; Skip is so
        MOV    AL,TIME_DAWN                         ; Reset Time Of day
        CALL   ChangeTimeOfDay                      ;
        MOV    CL,BYTE PTR 00Ah                     ; DAWN_ENERGY_BOOST_AI (10)
        CALL   GetCharLoyalty                       ; Is the character
        CMP    AL,RequiredLoyalty                   ; loyal?
        JZ     L662E                                ; Yes so Don't reduce as much.
        MOV    CL,BYTE PTR 014h                     ; DAWN_ENERGY_BOOST (20)
L662E:  CALL   GetCharTired                         ;
        CALL   Add_C_To_A                           ; Reduce tiredness
        CALL   PutCharTired                         ;
        CALL   GetCharFlag_1
        RES    CHARF_KILLED,AL                      ; Reset Killed
        RES    CHARF_INBATTLE,AL                    ; Reset In Battle
        RES    CHARF_WONBATTLE,AL                   ; Reset Won Battle
        CALL   PutCharFlag_1
        XOR    AL,AL
        CALL   PutArmySlew                          ; Reset Army Slew
        CALL   PutArmyLost                          ; Reset Army Lost
        MOV    AL,0ffh
        CALL   PutCharFightWho                      ; Reset Fighting Who
        CALL   MoveCharacterSomeWhere
L6652:  POPF   
        POP    AX
        INC    AL                                    ; Next Character
        CMP    AL,080h                               ; Have we done all 128
        JNZ    L6610                                 ; loop until we have

```

**Reply by FrozenEmpire** [reply]
In the LOM scenario the code does add energy based on time remaining… just in the DDR scenario it’s a fixed amount

but the main point I was raising is that dawn boosts are added before the night battles, so the characters benefit from the energy boosts during the nights battles. Given it’s a “Dawn boost” I guess I was expecting that the energy would get added at Dawn, ie after the nights battles…

**Reply by IcemarkUK** [reply]
Yes. I think again, this is a problem with my disassembly rather than Mike's original code. If you look at the original, it's not really a Dawn thing - it's an End of day thing... The Dawn terminology just comes from the fact that the original resets the time to dawn. From a gameplay POV I would agree that fighting at the end of the day, you would expect this to be on top of all the movement etc... but then are we fighting at the end of the day, or at the beginning of the next day? I guess horses for courses in a turn based game.

All that aside, the adjustments are made at the correct time in the cycle per the original.

**Reply by FrozenEmpire** [reply]
At least in my mind all the texts created a sense of battling during the night hours, before the next dawn

if the energy boost in the original was at the “end of the day” then I guess it should be kept like that, but doesn’t really make any sense to me gameplay wise. What causes that energy boost before battling… 

energy from resting is more clearly understandable

**Reply by IcemarkUK** [reply]
Starting to look like a gameplay option :)
